tree/ntuple/v7/inc/ROOT/RCluster.hxx

```diff
@@ -0,0 +1,183 @@
+
+// clang-format off
+/**
+\class ROOT::Experimental::Detail::ROnDiskPageMap
+\ingroup NTuple
+\brief A memory region that contains packed and compressed pages
+
+Derived classes implement how the on disk pages are stored in memory, e.g. mmap'd or in a special area.
+*/
+// clang-format on
+class ROnDiskPageMap {
+   friend class RCluster;
+
+protected:
+   /// The memory region containing the on-disk pages. Ownership of the memory region is passed to the page map.
+   /// Therefore, the region needs to be allocated in a way that fits the derived class and its destructor.
+   void *fMemory;
+   std::unordered_map<ROnDiskPage::Key, ROnDiskPage> fOnDiskPages;
+
+public:
+   explicit ROnDiskPageMap(void *memory) : fMemory(memory) {}
+   ROnDiskPageMap(const ROnDiskPageMap &other) = delete;
+   ROnDiskPageMap(ROnDiskPageMap &&other);
+   ROnDiskPageMap &operator =(const ROnDiskPageMap &other) = delete;
+   ROnDiskPageMap &operator =(ROnDiskPageMap &&other);
+   virtual ~ROnDiskPageMap();
+
+   /// Inserts information about a page stored in fMemory.  Therefore, the address referenced by onDiskPage
+   /// needs to be owned by the fMemory block.  If a page map contains a page of a given column, it is expected
+   /// that _all_ the pages of that column in that cluster are part of the page map.
+   void Register(const ROnDiskPage::Key &key, const ROnDiskPage &onDiskPage) { fOnDiskPages.emplace(key, onDiskPage); }
+};
+
+
+// clang-format off
+/**
+\class ROOT::Experimental::Detail::ROnDiskPageMapHeap
+\ingroup NTuple
+\brief An ROnDiskPageMap that is used for an fMemory allocated as an array of unsigned char.
+*/
+// clang-format on
+class ROnDiskPageMapHeap : public ROnDiskPageMap {
+public:
+   explicit ROnDiskPageMapHeap(void *memory) : ROnDiskPageMap(memory) {}
+   ~ROnDiskPageMapHeap();
+};
+
+// clang-format off
+/**
+\class ROOT::Experimental::Detail::RCluster
+\ingroup NTuple
+\brief An in-memory subset of the packed and compressed pages of a cluster
+
+Binds together several page maps that represent all the pages of certain columns of a cluster
+*/
+// clang-format on
+class RCluster {
+protected:
+   /// References the cluster identifier in the page source that created the cluster
+   DescriptorId_t fClusterId;
+   /// Multiple page maps can be combined in a single RCluster
+   std::vector<ROnDiskPageMap> fPageMaps;
+   /// List of the (complete) columns represented by the RCluster
+   std::unordered_set<DescriptorId_t> fAvailColumns;
+   /// Lookup table for the on-disk pages
+   std::unordered_map<ROnDiskPage::Key, ROnDiskPage> fOnDiskPages;
+
+public:
+   explicit RCluster(DescriptorId_t clusterId) : fClusterId(clusterId) {}
+   RCluster(const RCluster &other) = delete;
+   RCluster(RCluster &&other) = default;
+   RCluster &operator =(const RCluster &other) = delete;
+   RCluster &operator =(RCluster &&other) = default;
+   ~RCluster() = default;
+
+   /// Move the given page map into this cluster; overlapping page ranges are ignored
```

Comment:
What does "overlapping page ranges are ignored" means?

Reply:
Clarified in the code comment:

    on-disk pages that are present in both the cluster at hand and pageMap are gracefully 
    handled such that a following lookup will return the page from either of the 
    memory regions